README.md
<h1 align="center">
  <!-- Logo -->
  <a href="https://rill.site" alt="Rill">
    <img src="https://cdn.rawgit.com/rill-js/rill/master/Rill-Logo.svg" width="300" alt="Rill Logo"/>
  </a>

  <br/>

  <!-- Stability -->
  <a href="https://nodejs.org/api/documentation.html#documentation_stability_index">
    <img src="https://img.shields.io/badge/stability-stable-brightgreen.svg?style=flat-square" alt="API stability"/>
  </a>
  <!-- TypeScript -->
  <a href="http://typescriptlang.org">
    <img src="https://img.shields.io/badge/%3C%2F%3E-typescript-blue.svg" alt="TypeScript"/>
  </a>
  <!-- NPM version -->
  <a href="https://npmjs.org/package/rill">
    <img src="https://img.shields.io/npm/v/rill.svg?style=flat-square" alt="NPM version"/>
  </a>
  <!-- Travis build -->
  <a href="https://travis-ci.org/rill-js/rill">
  <img src="https://img.shields.io/travis/rill-js/rill.svg?style=flat-square" alt="Build status"/>
  </a>
  <!-- Coveralls coverage -->
  <a href="https://coveralls.io/github/rill-js/rill">
    <img src="https://img.shields.io/coveralls/rill-js/rill.svg?style=flat-square" alt="Test Coverage"/>
  </a>
  <!-- Downloads -->
  <a href="https://npmjs.org/package/rill">
    <img src="https://img.shields.io/npm/dm/rill.svg?style=flat-square" alt="Downloads"/>
  </a>
  <!-- Size -->
  <a href="https://npmjs.org/package/rill">
    <img src="https://img.shields.io/badge/gzipped-12.2kb-green.svg" alt="Browser Bundle Size (Gzipped)"/>
  </a>
  <!-- Gitter chat -->
  <a href="https://gitter.im/rill-js/rill">
    <img src="https://img.shields.io/gitter/room/rill-js/rill.svg?style=flat-square" alt="Gitter Chat"/>
  </a>
  <!-- Saucelabs -->
  <a href="https://saucelabs.com/u/rill-js">
    <img src="https://saucelabs.com/browser-matrix/rill-js.svg" alt="Sauce Test Status"/>
  </a>
</h1>

Expressive router for [nodejs](https://nodejs.org) and the browser.
Rill brings cascading middleware to the browser and enables a familiar routing solution for web applications.

Rill provides the minimum for abstractions over [nodejs](https://nodejs.org) and the browser enabling things like routing (with redirecting, refreshes and more), cookies, and middleware with the same api.

It supports many view engines including [Marko](http://markojs.com), [React](https://reactjs.org), [Svelte](https://svelte.technology) and even html only template engines such as [Pug](https://pugjs.org).

# Installation

```console
npm install rill
```

# Browser support
All modern browsers are supported including IE10 and above.
Older browsers will need to polyfill the Promise API, checkout [es6-promise](https://github.com/stefanpenner/es6-promise) for a good polyfill, babel-polyfill also covers this.

# Community

* [API Documentation](https://rill.site/application#docs)
* Examples
  * [React](https://github.com/rill-js/todomvc-react)
  * [Marko](https://github.com/rill-js/todomvc-marko)
  * [Svelte](https://github.com/rill-js/todomvc-svelte)
* [Wiki](https://github.com/rill-js/rill/wiki)
* [FAQ](https://github.com/rill-js/rill/wiki/FAQ)
* [Middleware List](https://github.com/rill-js/rill/wiki/Middleware)
* [Gitter Community](https://gitter.im/rill-js/rill)
* [Reddit Community](https://www.reddit.com/r/Rill)

# Articles

* [Universal web application framework - Interview with Dylan Piercey](https://survivejs.com/blog/rill-interview/)
* [Isomorphic Javascript, let’s make it easier.](https://medium.com/@pierceydylan/isomorphic-javascript-it-just-has-to-work-b9da5b0c8035)
* [How to make universal JavaScript applications — Part 1](https://medium.com/@pierceydylan/how-to-make-universal-javascript-applications-part-1-90e9032bc471)
* [Browsers, Servers, and APIs](https://medium.com/@iamjohnhenry/browsers-servers-and-apis-2f7b10523f39)
* [Why Everyone is Talking About Isomorphic](https://medium.com/capital-one-developers/why-everyone-is-talking-about-isomorphic-universal-javascript-and-why-it-matters-38c07c87905#.mdd84j28m)
* [Isomorphic JavaScript: The Future of Web Apps](https://medium.com/airbnb-engineering/isomorphic-javascript-the-future-of-web-apps-10882b7a2ebc)

# Why Rill?
Rill is the answer to a simple question; Can I run my [Express](https://github.com/expressjs/express) style router in the browser? Turns out you can and it works awesome.

It brings a common interface to many typical app like features in both the browser and [nodejs](https://nodejs.org). Many isomorphic frameworks and routers have crazy abstractions and learning curves but with Rill, if you understand [Express](https://github.com/expressjs/express) or [Koa](https://github.com/koajs/koa), you already know how the routing works! In Rill you get to program much of your application logic using the same api (client or server) including routing, rendering, data fetching and more are easily shared.

Rill also works perfectly as a stand alone router for [nodejs](https://nodejs.org) or in the browser. This allows for easy progressive enhancement. If all is well the browser can handle much of your application logic and if JavaScript fails for any reason your server knows exactly what to do.

# How does this thing work?
If you look at the source for Rill [here](https://github.com/rill-js/rill/tree/master/src) you will quickly notice there is ZERO browser specific code. This is all thanks to [@rill/http](https://github.com/rill-js/http) which is node's [HTTP.createServer](https://nodejs.org/api/http.html#http_http_createserver_requestlistener) ported to the browser.

In the browser it works by listening for internal link clicks, form submissions and browser history changes. It will then create a [Rill Context](https://rill.site/context#docs) for each of these events and emit it through the router, similar to how receiving a request works in [nodejs](https://nodejs.org).

It supports everything you'd expect from a client side [nodejs](https://nodejs.org) server. This includes redirects, refreshes, cookies, scrolling and url updates using the [History API](https://developer.mozilla.org/en-US/docs/Web/API/History).

# Example

### Create an app

```javascript
/**
 * The following code can run 100% in the browser or in nodejs.
 * Examples use es2015/2016 with Babel and JSX but this is optional.
 */

import Rill from 'rill'
const app = new Rill() // You can call Rill without new, but autocomplete will not work.
```

### Setup middleware

```javascript
// Universal form data parsing middleware.
import bodyParser from '@rill/body'
app.use(bodyParser())

// Universal react rendering middleware.
import reactRenderer from '@rill/react'
app.use(reactRenderer())

// Example Logger
app.use(async ({ req }, next)=> {
  const start = Date.now()

  // Rill uses promises for control flow.
  // ES2016 async functions work great as well!
  await next()

  const ms = Date.now() - start
  console.log(`${req.method} ${req.url} - ${ms}`)
})
```

### Setup a page

```javascript
// Respond to a GET request.
app.get('/todos', async ({ res })=> {
  // Fetch a todolist from some service.
  const todolist = await MyTodoListService.getAllTodos()

  // Directly set React virtual dom to the body thanks to @rill/react.
  // (Checkout @rill/html for universal html diffing).
  res.body = (
    <html>
      <head>
        <title>My App</title>
        <meta name="description" content="Rill Application">
      </head>
      <body>
        <form action="/add-todo" method="POST">
          <h1>Just a plain old form</h1>
          <input type="text" name="todo"/>
          <button type="submit">Add Todo</button>
        </form>

        {todolist.length
          ? todolist.map(renderTodo)
          : 'No todos to display.'
        }
        <script src="/app.js"/>
      </body>
    </html>
  )
})
```

### Handle a form submission
```javascript
// Respond to a POST request.
app.post('/add-todo', async ({ req, res })=> {
  // We handle form submissions with Rill the same way one would express or koa.
  // Here we are simply adding the todo via some service.
  await MyTodoListService.addTodo({ text: req.body.todo })
  // And then we redirect back (same as res.redirect('/todos'))
  res.redirect('back')
})
```

### Start app

```javascript
// Start a regular http server.
// In the browser any form submissions or link clicks will intercepted by @rill/http.
app.listen({ port: 80 })
```

## See Also

* [isbrowser](https://github.com/DylanPiercey/isbrowser) - A browserify transform to remove server-side code.
* [isomorphic-fetch](https://github.com/matthew-andrews/isomorphic-fetch) - Universal http requests using WHATWG fetch.
* [isomorphic-form-data](https://github.com/form-data/isomorphic-form-data) - Send multipart form data universally (able to send files and works with fetch).
* [scroll-behavior](https://github.com/DylanPiercey/scroll-behaviour) - @rill/http will automatically try to use the "smooth" [scroll-behavior](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-behavior) when scrolling to targets on link clicks. This will polyfill that across modern browsers.
* [submit-form](https://github.com/DylanPiercey/submit-form) - Manually trigger Rill navigation in the browser.

## Prior Art

* [koa-client](https://github.com/kentjs/koa-client) - Koa clone that runs in the browser, inspired this package.
* [monorouter](https://github.com/matthewwithanm/monorouter) - Another isomorphic router that partially inspired this package.

## Contributions

* Use `npm test` to build and run tests.

## License

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

# 7.0.3, 7.0.4 - 2018-02-27
- Add type definitions for params, querystring, cookies and cookie serialization options.
- Add type definitions for TLS options.
- Updated dev dependencies.
- Update license dates.

# 7.0.1, 7.0.2 - 2017-11-18
- Update links to todomvc examples in readme.
- Fix bad URL for size badge and update size in readme.

# 7.0.0 - 2017-11-17
- Changes to handling URLS with a `hash` in @rill/http.
  * No longer intercept `hash` changes (allows for default browser action).
  * No longer add `hash` to `req.url` or `req.headers.referer` for consistency with server side http.
  * Removes built in smooth scrolling on urls with a `hash`.
- `res.status` now starts as `undefined` instead of `404`. (Will still default to 404 if no body set during response).
- `ctx.fail` and `ctx.assert` no longer set `res.status` or `res.message` if the error was caught.
- Drop official support for IE9.
- Remove official bower support.
- No longer provide prebuilt bundle.
- Fully convert project to typescript. (Autocomplete for VSCode!)
- Setup tslint and prettier for formatting.

# 6.0.0, 6.0.1, 6.0.2 - 2017-10-10
- Upgraded @rill/http to 6.0.0
- Better support for typescript with improved JSDoc.
- Sets booleans by default instead of undefined for `res.finished`, `res.end` and `res.respond`.

# 5.0.13, 5.0.14 - 2017-08-03
- Remove [obsolete version field](https://github.com/bower/spec/blob/master/json.md#version) from bowser.json.
- Add related airbnb article to readme.
- Updated dev dependencies.

## 5.0.10, 5.0.11, 5.0.12 - 2017-07-30
- Updated `middleware list` url in readme.
- Add benchmarks folder for comparisons with Koa/Express.
- Switch uglifyjs back to dev dependency.

## 5.0.7, 5.0.8, 5.0.9 - 2017-07-27
- Switch to white listing files in package.json.
- Added survivejs interview.
- Added bundle size analyzer script for future improvements.
- Update dependencies.
- Update license year.
- Update bower version.

## 5.0.6 - 2017-07-19
- Switched logo to SVG.

## 5.0.4, 5.0.5 - 2017-06-13
- Added new article.
- Switch mocha tests to use promises.

## 5.0.1, 5.0.2, 5.0.3 - 2017-06-06
- Update dependencies.
- No longer relies on `process.browser`, instead uses browser field to attach to document.

## 5.0.0 - 2017-05-18
- Upgraded @rill/http which has a [breaking change](https://github.com/rill-js/http/blob/master/CHANGELOG.md#500-501---2017-05-18). (This change will only be breaking for those who rely on @rill/http directly)

## 4.3.4, 4.3.5 - 2017-04-23
- Update dependencies.
- Misc documentation cleanup.

## 4.3.2, 4.3.3 - 2017-03-28
- Update dependencies. (Fixes issue in mini-url where the base URL was not optional in some browsers).

## 4.3.1 - 2017-03-23
- Fix download link in readme.

## 4.3.0 - 2017-03-23
- Ensure all requests in the browser are async.


## 4.2.0, 4.2.1 - 2017-03-23
- Update @rill/http which added new options to the browser adapter's 'fetch' method. Also a url string is now allowed as the first argument.

## 4.1.5 - 2017-03-23
- Fix regression with parsing querystrings on GET forms.

## 4.1.3, 4.1.4 - 2017-03-21
- Update logo link for npmjs.com.
- More JSDoc improvements.

## 4.1.2 - 2017-03-17
- Fixed regression where referrer header was no longer the full href.

## 4.1.1 - 2017-03-14
- Improve inline JSDoc comments.

## 4.1.0 - 2017-03-11
- Removed need for fetch api polyfill. (Promise polyfill still needed for older browsers).
- Optimized Rill's file size in several ways.
  * Switched to [events-light](https://www.npmjs.com/package/events-light) from browserify events shim.
  * Switch to [mini-url](https://www.npmjs.com/package/mini-url) and [mini-querystring](https://www.npmjs.com/package/mini-querystring) from browserify url and querystring shims.
  * Removed unused methods from @rill/http.
  * Removed Buffer dependency in the browser, now uses [Blobs](https://developer.mozilla.org/en/docs/Web/API/Blob).

## 3.0.0 - 2016-11-01
- Updated to majorly refactored and fully tested @rill/http.
  * Added Rill#createServer method to get a node server from the rill app.
  * Switch to [https://github.com/jshttp/cookie](cookie) for cookie parsing.
  * Other non breaking changes to [@rill/http changelogs](https://github.com/rill-js/http/blob/master/CHANGELOG.md#user-content-300---2016-10-31).

## 2.5.2 - 2016-11-08
- Add node 7.0.0 to travis build.

## 2.5.0, 2.5.1 - 2016-11-08
- Fix for breaking change in HttpServer#listen in node 7.0.0.

## 2.4.0 - 2016-10-12
- Updated @rill/http. This fixes an issue in safari where named submit buttons were not being parsed on form submissions.
- Updated pathToRegexp. This allows for better hostname matching support (previously some issues existed for splat host names such as 'mysite.:something+.com'), this should now be fixed.

## 2.3.0 - 2016-10-01
- Exposes `ctx.req.matchPath` and `ctx.req.matchHost` which allow users to check what the current `path` and `host` being matched are.
- Splat paths (such as `/test/*`) will now properly make trailing slashes optional. In this case `/test` is now treated as valid where before only `/test/` would have been valid.

## 2.1.8 - 2016-09-28
- Updated @rill/error which now has two new methods (fail and assert).

## 2.1.7 - 2016-09-24
- Added another isomorphic article to README.
- Updated dev dependencies.

## 2.1.2, 2.1.3, 2.1.4, 2.1.5, 2.1.6 - 2016-09-17
- Link rill logo to [rill website](https://rill.site).
- Changed readme logo to use SVG from [rill website](https://rill.site).
- Updated any references to docs to the [rill website](https://rill.site).
- Move some NPM scripts to makefile.
- Remove duplicate info in docs.
- Remove some unneeded files when installing with bower.
### Added
- Added docs back to package.json files for use on the docs website.


src/index.js
// @ts-check
/** Type Definitions */
/** @module rill */
/** @typedef {(app: Rill) => any} TransformFunction */
/** @typedef {TransformFunction|false|void} TransformArg */
/** @typedef {(ctx: Context, next: function?) => any} MiddlewareFunction */
/** @typedef {MiddlewareFunction|Rill|false|void} MiddlewareArg */
/**
 * @typedef {object} listenOptions
 * @property {string} [ip] - The interface to listen on.
 * @property {number} [port] - The port to listen on.
 * @property {number} [backlog] - The maximum length of the queue of pending connections.
 * @property {object} [tls] - Accepts options from tls.createServer() and tls.createSecureContext().
 */
'use strict'

var pathToRegExp = require('path-to-regexp')
var http = require('@rill/http')
var https = require('@rill/https')
var chain = require('@rill/chain')
var HttpError = require('@rill/error')
var Context = require('./context')
var respond = require('./respond')
/* istanbul ignore next */
var attachDocument = require('./attach')
var parse = pathToRegExp.parse
var tokensToRegExp = pathToRegExp.tokensToRegExp
var slice = Array.prototype.slice
module.exports = Rill['default'] = Rill

/**
 * Creates a universal app that will run middleware for a incoming request.
 *
 * @example
 * const app = Rill()
 *
 * @constructor
 */
function Rill () {
  if (!(this instanceof Rill)) return new Rill()
  /** @type {MiddlewareArg[]} */
  this.stack = []
}

/**
 * Takes the current middleware stack, chains it together and
 * returns a valid handler for a node js style server request.
 *
 * @example
 * const app = Rill()
 * app.use(...)
 * require('http').createServer(app.handler()).listen()
 *
 * @return {(req: http.IncomingMessage, res: http.ServerResponse) => void}
 */
Rill.prototype.handler = function () {
  var fn = chain(this.stack)

  /**
   * Handles a node js server request and pushes it through a rill server.
   *
   * @param {http.IncomingMessage} req - The http request.
   * @param {http.ServerResponse} res - The http response.
   * @return {void}
   */
  return function handleIncomingMessage (req, res) {
    res.statusCode = 404
    var ctx = new Context(req, res)

    fn(ctx)
      .catch(handleError)
      .then(finish)

    function handleError (err) {
      if (Number(ctx.res.status) === 404) {
        ctx.res.status = 500
      }
      if (!(err instanceof HttpError)) {
        /* istanbul ignore next */
        console && console.error && console.error(err)
      }
    }

    function finish () { respond(ctx) }
  }
}

/**
 * Creates a node server from the Rill server.
 *
 * @example
 * app.createServer().listen(3000)
 *
 * @param {object} [tls] - Node https TLS options.
 * @return {http.Server}
 */
Rill.prototype.createServer = function (tls) {
  var handler = this.handler()
  var server = tls ? https.createServer(tls, handler) : http.createServer(handler)
  // Setup link hijacking in the browser.
  /* istanbul ignore next */
  attachDocument(server)
  return server
}

/**
 * Creates a node server from the current Rill server and starts listening for http requests.
 *
 * @example
 * rill().use(...).listen({ port: 3000 })
 *
 * @param {listenOptions} [options] - Options to configure the node server.
 * @param {function} [onListening] - A function to be called once the server is listening for requests.
 * @returns {http.Server}
 */
Rill.prototype.listen = function (options, onListening) {
  // Make options optional.
  if (typeof options === 'function') {
    onListening = options
    options = null
  }

  options = options || {}
  options.port = options.port != null ? options.port : 0
  return this.createServer(options.tls).listen(options.port, options.ip, options.backlog, onListening)
}

/**
 * Append new middleware to the current rill application stack.
 *
 * @example
 * rill.use(fn1, fn2)
 *
 * @param {...MiddlewareArg} [middleware] - A middleware to attach.
 * @return {Rill}
 */
Rill.prototype.use = function (middleware) {
  var start = this.stack.length
  var end = this.stack.length += arguments.length

  for (var i = end; start < i--;) {
    this.stack[i] = arguments[i - start]
  }

  return this
}

/**
 * Simple syntactic sugar for functions that
 * wish to modify the current rill instance.
 *
 * @example
 * app.setup(self => {
 *  // Modify the current app.
 *  self.use(...)
 *  self.modified = true
 * })
 *
 * @param {...TransformArg} transformer - A function that will modify the rill instance.
 * @return {Rill}
 */
Rill.prototype.setup = function (transformer) {
  for (var fn, len = arguments.length, i = 0; i < len; i++) {
    fn = arguments[i]
    if (!fn) continue
    if (typeof fn === 'function') fn(this)
    else throw new TypeError('Rill#setup: Setup must be a function or falsey.')
  }

  return this
}

/**
 * Use middleware at a specific pathname.
 *
 * @example
 * app.at('/test', (ctx, next) => ...)
 *
 * @param {string} pathname - The pathname to match.
 * @param {...MiddlewareArg} [middleware] - A middleware to attach.
 * @return {Rill}
 */
Rill.prototype.at = function (pathname, middleware) {
  if (typeof pathname !== 'string') throw new TypeError('Rill#at: Path name must be a string.')

  var keys = []
  var reg = toReg(pathname, keys, { end: false, delimiter: '/' })
  var fn = chain(slice.call(arguments, 1))

  return this.use(function matchPathname (ctx, next) {
    var pathname = ctx.req.matchPath
    var matches = pathname.match(reg)
    // Check if we matched the whole path.
    if (!matches || matches[0] !== pathname) return next()

    // Check if params match.
    for (var key, match, i = keys.length; i--;) {
      key = keys[i]
      match = matches[i + 1]
      if (key.repeat) match = match == null ? [] : match.split('/')
      ctx.req.params[key.name] = match
    }

    // Update path for nested routes.
    var matched = matches[matches.length - 1] || ''
    if (ctx.req.matchPath !== matched) ctx.req.matchPath = '/' + matched

    // Run middleware.
    return fn(ctx, function () {
      // Reset nested pathname before calling later middleware.
      ctx.req.matchPath = pathname
      // Run sibling middleware.
      return next()
    })
  })
}

/**
 * Use middleware at a specific hostname.
 *
 * @example
 * app.host('test.com', (ctx, next) => ...)
 *
 * @param {string} hostname - The hostname to match.
 * @param {...MiddlewareArg} [middleware] - A middleware to attach.
 * @return {Rill}
 */
Rill.prototype.host = function host (hostname, middleware) {
  if (typeof hostname !== 'string') throw new TypeError('Rill#host: Host name must be a string.')

  var keys = []
  var reg = toReg(hostname, keys, { strict: true, delimiter: '.' })
  var fn = chain(slice.call(arguments, 1))

  return this.use(function matchHost (ctx, next) {
    var hostname = ctx.req.matchHost
    var matches = hostname.match(reg)

    // Check if we matched the whole hostname.
    if (!matches || matches[0] !== hostname) return next()

    // Here we check for the dynamically matched subdomains.
    for (var key, match, i = keys.length; i--;) {
      key = keys[i]
      match = matches[i + 1]
      if (key.repeat) match = match == null ? [] : match.split('.')
      ctx.req.subdomains[key.name] = match
    }

    // Update hostname for nested routes.
    var matched = matches[matches.length - 1] || ''
    if (ctx.req.matchHost !== matched) ctx.req.matchHost = matched

    // Run middleware.
    return fn(ctx, function () {
      // Reset nested hostname.
      ctx.req.matchHost = hostname
      // Run sibling middleware.
      return next()
    })
  })
}

// Attach all http verbs as shortcut methods.
http.METHODS.forEach(function (method) {
  var name = method.toLowerCase()
  /**
   * Use middleware on |method| requests at an (optional) pathname.
   *
   * @example
   * app.|method|('/test', ...)
   *
   * @param {string} [pathname] - A pathname to match.
   * @param {...MiddlewareArg} [middleware] - A middleware to attach.
   * @return {Rill}
   */
  Rill.prototype[name] = function (pathname, middleware) {
    var offset = typeof pathname === 'string' ? 1 : 0
    var fn = chain(slice.call(arguments, offset))
    if (offset === 1) return this.at(pathname, matchMethod)
    return this.use(matchMethod)

    function matchMethod (ctx, next) {
      if (ctx.req.method !== method) return next()
      return fn(ctx, next)
    }
  }
})

/**
 * Small wrapper around path to regexp that treats a splat param "/*" as optional.
 * This makes mounting easier since typically when you do a path like "/test/*" you also want to treat "/test" as valid.
 *
 * @param {string} pathname - The path to convert to a regexp.
 * @param {array} [keys] - A place to store matched param keys.
 * @param {object} [options] - Options passed to pathToRegExp.
 * @return {RegExp}
 */
function toReg (pathname, keys, options) {
  // First parse path into tokens.
  var tokens = parse(pathname)

  // Find the last token (checking for splat params).
  /** @type {object} */
  var splat = tokens[tokens.length - 1]

  // Check if the last token is a splat and make it optional.
  if (splat && splat.asterisk) splat.optional = true

  // Convert the tokens to a regexp.
  var re = tokensToRegExp(tokens, options)

  // Assign keys to from regexp.
  re.keys = keys
  for (var i = 0, len = tokens.length; i < len; i++) {
    if (typeof tokens[i] === 'object') keys.push(tokens[i])
  }

  return re
}
